store/boltstore/bolt_store.go

```diff
@@ -0,0 +1,484 @@
+
+	if val == nil {
+		return nil, ErrKeyNotFound
+	}
+	return append([]byte(nil), val...), nil
+}
+
+// SetUint64 is like Set, but handles uint64 values
+func (b *BoltStore) SetUint64(key []byte, val uint64) error {
+	return b.Set(key, uint64ToBytes(val))
+}
+
+// GetUint64 is like Get, but handles uint64 values
+func (b *BoltStore) GetUint64(key []byte) (uint64, error) {
+	val, err := b.Get(key)
+	if err != nil {
+		return 0, err
+	}
+	return bytesToUint64(val), nil
+}
+
+// encodePeers is used to serialize a Configuration into the old peers format.
+// This is here for backwards compatibility when operating with a mix of old
+// servers and should be removed once we deprecate support for protocol version 1.
+func encodePeers(configuration raft.Configuration, trans raft.Transport) []byte {
+	// Gather up all the voters, other suffrage types are not supported by
+	// this data format.
+	var encPeers [][]byte
+	for _, server := range configuration.Servers {
+		if server.Suffrage == raft.Voter {
+			encPeers = append(encPeers, trans.EncodePeer(server.ID, server.Address))
+		}
+	}
+
+	// Encode the entire array.
+	buf, err := encodeMsgPack(encPeers)
+	if err != nil {
+		panic(fmt.Errorf("failed to encode peers: %v", err))
+	}
+
+	return buf.Bytes()
+}
+
+// Create is used to start a new bbolt snapshot
+func (b *BoltStore) Create(version raft.SnapshotVersion, index, term uint64, configuration raft.Configuration,
+	configurationIndex uint64, trans raft.Transport) (raft.SnapshotSink, error) {
+	if version != 1 {
+		return nil, fmt.Errorf("unsupported snapshot version %d", version)
+	}
+
+	// Get the name of the snapshot
+	now := time.Now()
+	msec := now.UnixNano() / int64(time.Millisecond)
+	name := fmt.Sprintf("%d-%d-%d", term, index, msec)
+
+	sink := &BoltSnapshotSink{
+		boltStore: b,
+		meta: raft.SnapshotMeta{
+			Version:            version,
+			ID:                 name,
+			Index:              index,
+			Term:               term,
+			Peers:              encodePeers(configuration, trans),
+			Configuration:      configuration,
+			ConfigurationIndex: configurationIndex,
+		},
+		contents: &bytes.Buffer{},
+	}
+
+	return sink, nil
+}
+
+// getSnapshot returns a list of snapshot metadata in bbolt store
+func (b *BoltStore) getSnapshot(tx *bolt.Tx) []*raft.SnapshotMeta {
+	var snapMeta []*raft.SnapshotMeta
+	curs := tx.Bucket(dbSnapMeta).Cursor()
+	for k, v := curs.First(); k != nil; k, v = curs.Next() {
+		tempMeta := new(raft.SnapshotMeta)
+		if decodeMsgPack(v, tempMeta) == nil {
+			snapMeta = append(snapMeta, tempMeta)
+		}
+	}
+
+	// sort the snapshot in reverse order (new->old)
+	sort.Sort(sort.Reverse(snapMetaSlice(snapMeta)))
+	return snapMeta
+}
+
+// List returns available snapshots in bbolt store.
+func (b *BoltStore) List() ([]*raft.SnapshotMeta, error) {
+	tx, err := b.conn.Begin(false)
+	if err != nil {
+		return []*raft.SnapshotMeta{}, err
+	}
+	defer tx.Rollback()
+
+	snapMeta := b.getSnapshot(tx)
+	if len(snapMeta) > b.retain {
```

Comment:
Yeah, but if the order matters?

Reply:
Use the `curs.Last()` you can get the latest snapshots.

Reply:
Because we put the key-value pairs into bbolt only when calling ```Close```. The comparison is as follows. What if a user calls ```Close``` with the newest snapshot first, and then the second newest one? In this case, ```cur.Last()``` does not return the newest snapshot.

Reply:
Sounds great.